### PR TITLE
[GraphBolt] Add sampling graph C side code

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
This is the C side code of csr sampling graph, python code and related binding and test will be in anohter PR after this is merged.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by Rhett-Ying — COMMENTED** [review]
could you enable the cpp unit tests for graphbolt? probably it's better to enable in another standalone PR.

File: graphbolt/include/csr_sampling_graph.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  int64_t NumCols() const { return num_cols_; }
+
+  /** @brief Get the number of nodes. */
+  int64_t NumNodes() const { return indptr_.size(0) - 1; }
+
+  /** @brief Get the number of edges. */
+  int64_t NumEdges() const { return indices_.size(0); }
+
+  /** @brief Get the index pointer tensor. */
+  const torch::Tensor IndPtr() const { return indptr_; }
```
Comment (czkkkkkk) [inline graphbolt/include/csr_sampling_graph.h:93]: Maybe better to use `CSCIndptr` to explicitly say the storage is in CSC format.

File: graphbolt/include/csr_sampling_graph.h [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+using StringList = std::vector<std::string>;
+
+/**
+ * @brief Heterogeneous information in the Graph.
+ */
+struct HeteroInfo {
+  HeteroInfo(
+      const StringList& ntypes, const StringList& etypes,
+      torch::Tensor& node_type_offset_, torch::Tensor& type_per_edge_)
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csr_sampling_graph.h:24]: Fixed

File: graphbolt/include/csr_sampling_graph.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csr_sampling_graph.h:93]: fixed

File: graphbolt/include/csr_sampling_graph.h [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        edge_types(etypes),
+        node_type_offset(node_type_offset_),
+        type_per_edge(type_per_edge_) {}
+  /** @brief List of node types in the graph. */
+  StringList node_types;
+  /** @brief List of edge types in the graph. */
+  StringList edge_types;
+  /** @brief Offset array of node type. */
+  torch::Tensor node_type_offset;
+  /** @brief Type of each edge. */
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csr_sampling_graph.h:35]: Good suggestion, added for clarify.

File: graphbolt/include/csr_sampling_graph.h [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ */
+class CSRSamplingGraph : public torch::CustomClassHolder {
+ public:
+  /**
+   * @brief Constructor for CSR with data.
+   * @param num_rows The number of rows in the dense shape.
+   * @param num_cols The number of columns in the dense shape.
+   * @param indptr The CSR format index pointer array.
+   * @param indices The CSR format index array.
+   * @param hetero_info Heterogeneous graph information, if present.
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csr_sampling_graph.h:50]: Means the graph is homogeneous.

File: graphbolt/include/csr_sampling_graph.h (hunk 1 above)
Comment (peizhou001) [reply graphbolt/include/csr_sampling_graph.h:93]: Make sense, change for fix.

File: graphbolt/include/csr_sampling_graph.h [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+  /**
+   * @brief Set the heterogeneous information for the graph.
+   *
+   * @param ntypes A list of node types.
+   * @param etypes A list of edge types.
+   * @param node_type_offset_ A tensor representing the offset of node types.
+   * @param type_per_edge_ A tensor representing the type of each edge.
+   */
+  void SetHeteroInfo(
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csr_sampling_graph.h:76]: Make sense, it is originally used for test, let me remove it.

File: graphbolt/include/csr_sampling_graph.h [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      const StringList& ntypes, const StringList& etypes,
+      torch::Tensor node_type_offset_, torch::Tensor type_per_edge_);
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csr_sampling_graph.h:78]: fixed

**Comment by peizhou001** [comment]
> could you enable the cpp unit tests for graphbolt? probably it's better to enable in another standalone PR.

As we discussed
1. Python test can cover all scenarios.
2. C test is pretty time consuming and require CMake file change, let's not add this for now.

File: graphbolt/include/csr_sampling_graph.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply graphbolt/include/csr_sampling_graph.h]: Yes. I think we need to clarify this in the comment.

File: graphbolt/include/csr_sampling_graph.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply graphbolt/include/csr_sampling_graph.h]: I see. For API simplicity, we may have two interfaces for SamplingGraph construction in the Python side: `from_homo_csc()` and `from_hetero_csc()`. The first one is called without heterogeneous information while the latter one also takes the arguments in the current `SetHeteroInfo` as input. Anyway, I don't like `SetHeteroInfo` because it implies that the heterogeneous of a graph is mutable.

File: graphbolt/include/csr_sampling_graph.h (hunk 1 above)
Comment (czkkkkkk) [reply graphbolt/include/csr_sampling_graph.h]: > class name is already starts with CSC, so users should know it is a CSC graph

Make sense. 

> I may perfer IndPtr, one is to keep consistent with other member variables naming convention.

To me, `indptr` is more like a special word for CSR format. And if you use `IndPtr`, you may also need to use `ind_ptr_`. That is a bit weird... Anyway, I am not strongly insist on this.

File: graphbolt/include/csr_sampling_graph.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csr_sampling_graph.h]: Make sense, added for fix.

File: graphbolt/include/csr_sampling_graph.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csr_sampling_graph.h]: Make sense, chang for fix.

File: graphbolt/CMakeLists.txt [hunk 7]
```diff
@@ -23,10 +23,6 @@ list(GET TORCH_PREFIX_VER 1 TORCH_VER)
 message(STATUS "Configuring for PyTorch ${TORCH_VER}")
 string(REPLACE "." ";" TORCH_VERSION_LIST ${TORCH_VER})
 
-if(USE_CUDA)
```
Comment (frozenbugs) [inline graphbolt/CMakeLists.txt:26]: Split out to a different PR, unrelated to your code.

Comment (peizhou001) [reply graphbolt/CMakeLists.txt:26]: sure, remove for fix

**Review by czkkkkkk — APPROVED** [review]

File: graphbolt/include/csc_sampling_graph.h [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+#include <string>
+#include <vector>
+
+namespace graphbolt {
+namespace sampling {
+
+using StringList = std::vector<std::string>;
+
+/**
+ * @brief Heterogeneous information in the Graph.
```
Comment (jermainewang) [inline graphbolt/include/csc_sampling_graph.h:19]: Need more docstring. Please describe (e.g., using an example) what's the concrete semantics of each member.

File: graphbolt/include/csc_sampling_graph.h [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+   * @param indices The CSC format index array.
+   * @param hetero_info Heterogeneous graph information, if present. Nullptr
+   * means it is a homogeneous graph.
+   */
+  CSCSamplingGraph(
+      int64_t num_rows, int64_t num_cols, torch::Tensor& indptr,
+      torch::Tensor& indices, const std::shared_ptr<HeteroInfo>& hetero_info);
+
+  /**
+   * @brief Create a CSC graph from tensors of CSC format.
```
Comment (jermainewang) [inline graphbolt/include/csc_sampling_graph.h:59]: Is this for constructing homogeneous graph? If so, please state this clearly.

File: graphbolt/include/csc_sampling_graph.h [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+   */
+  static c10::intrusive_ptr<CSCSamplingGraph> FromCSC(
+      const std::vector<int64_t>& shape, torch::Tensor indptr,
+      torch::Tensor indices);
+
+  /**
+   * @brief Create a CSC graph from tensors of CSC format.
+   * @param indptr Index pointer array of the CSC.
+   * @param indices Indices array of the CSC.
+   * @param shape Shape of the sparse matrix.
```
Comment (jermainewang) [inline graphbolt/include/csc_sampling_graph.h:74]: strange docstring. It is for graph construction, what is "sparse matrix"?

File: graphbolt/include/csc_sampling_graph.h [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+   */
+  CSCSamplingGraph(
+      int64_t num_rows, int64_t num_cols, torch::Tensor& indptr,
+      torch::Tensor& indices, const std::shared_ptr<HeteroInfo>& hetero_info);
+
+  /**
+   * @brief Create a CSC graph from tensors of CSC format.
+   * @param indptr Index pointer array of the CSC.
+   * @param indices Indices array of the CSC.
+   * @param shape Shape of the sparse matrix.
```
Comment (jermainewang) [inline graphbolt/include/csc_sampling_graph.h:62]: strange docstring. It is for graph construction, what is "sparse matrix"?

File: graphbolt/include/csc_sampling_graph.h [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+   *
+   * @return CSCSamplingGraph
+   */
+  static c10::intrusive_ptr<CSCSamplingGraph> FromCSC(
+      const std::vector<int64_t>& shape, torch::Tensor indptr,
```
Comment (jermainewang) [inline graphbolt/include/csc_sampling_graph.h:67]: Why this interface accepts a vector<int> while some of the previous one accepts two ints `num_rows` and `num_cols`?

File: graphbolt/include/csc_sampling_graph.h [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+      torch::Tensor node_type_offset, torch::Tensor type_per_edge);
+
+  /** @brief Get the number of rows. */
+  int64_t NumRows() const { return num_rows_; }
+
+  /** @brief Get the number of columns. */
+  int64_t NumCols() const { return num_cols_; }
+
+  /** @brief Get the number of nodes. */
+  int64_t NumNodes() const { return indptr_.size(0) - 1; }
```
Comment (jermainewang) [inline graphbolt/include/csc_sampling_graph.h:96]: Does `CSCSamplingGraph` accept `num_rows` != `num_cols`? Shall we further simplify it?

File: graphbolt/include/csc_sampling_graph.h [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+  /** @brief Get the number of edges. */
+  int64_t NumEdges() const { return indices_.size(0); }
+
+  /** @brief Get the csc index pointer tensor. */
+  const torch::Tensor CSCIndptr() const { return indptr_; }
```
Comment (jermainewang) [inline graphbolt/include/csc_sampling_graph.h:102]: Why `CSCIndptr` instead of `Indptr`?

File: graphbolt/include/csc_sampling_graph.h [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+  /** @brief Get the list of node types for a heterogeneous graph. */
+  inline StringList& NodeTypes() const { return hetero_info_->node_types; }
+
+  /** @brief Get the list of edge types for a heterogeneous graph. */
+  inline const StringList& EdgeTypes() const {
+    return hetero_info_->edge_types;
+  }
+
+  /** @brief Get the edge type tensor for a heterogeneous graph. */
+  inline const torch::Tensor PerEdgeType() const {
```
Comment (jermainewang) [inline graphbolt/include/csc_sampling_graph.h:124]: Why `PerEdgeType` instead of `TypePerEdge`?

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: graphbolt/include/csc_sampling_graph.h [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        type_per_edge(type_per_edge) {}
+  /** @brief List of node types in the graph. */
+  StringList node_types;
+  /** @brief List of edge types in the graph. */
+  StringList edge_types;
+  /** @brief Offset array of node type. */
+  torch::Tensor node_type_offset;
+  /** @brief Type id of each edge, where type id is the corresponding index of
+   * edge_types. */
+  torch::Tensor type_per_edge;
```
Comment (frozenbugs) [inline graphbolt/include/csc_sampling_graph.h:37]: ditto

File: graphbolt/include/csc_sampling_graph.h (hunk 8 above)
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:19]: added

File: graphbolt/include/csc_sampling_graph.h (hunk 9 above)
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:59]: Yes, add description in docstring for clarify.

File: graphbolt/include/csc_sampling_graph.h (hunk 10 above)
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:74]: changed to graph for fix.

File: graphbolt/include/csc_sampling_graph.h (hunk 13 above)
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:96]: Yes, we support it, looks it won't introduce more complexity to our implementation, so maybe just keep it here.

File: graphbolt/include/csc_sampling_graph.h (hunk 14 above)
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:102]: According to @czkkkkkk 's suggestion, `Indptr` is de-facto specilized for Csr structure, so use `CSCIndptr ` to distinguish them.

File: graphbolt/include/csc_sampling_graph.h (hunk 15 above)
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:124]: It should be a mistake, changed to TypePerEdge for consistentence.

File: graphbolt/include/csc_sampling_graph.h [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+  inline StringList& NodeTypes() const { return hetero_info_->node_types; }
+
+  /** @brief Get the list of edge types for a heterogeneous graph. */
+  inline const StringList& EdgeTypes() const {
+    return hetero_info_->edge_types;
+  }
+
+  /** @brief Get the edge type tensor for a heterogeneous graph. */
+  inline const torch::Tensor PerEdgeType() const {
+    return hetero_info_->type_per_edge;
```
Comment (frozenbugs) [inline graphbolt/include/csc_sampling_graph.h:125]: Make it consistent, both per_edge_type or both TypePerEdge.

File: graphbolt/include/csc_sampling_graph.h [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+        node_type_offset(node_type_offset),
+        type_per_edge(type_per_edge) {}
+  /** @brief List of node types in the graph. */
+  StringList node_types;
+  /** @brief List of edge types in the graph. */
+  StringList edge_types;
+  /** @brief Offset array of node type. */
+  torch::Tensor node_type_offset;
+  /** @brief Type id of each edge, where type id is the corresponding index of
+   * edge_types. */
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:57]: changed

File: graphbolt/include/csc_sampling_graph.h [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+      : node_types(ntypes),
+        edge_types(etypes),
+        node_type_offset(node_type_offset),
+        type_per_edge(type_per_edge) {}
+  /** @brief List of node types in the graph. */
+  StringList node_types;
+  /** @brief List of edge types in the graph. */
+  StringList edge_types;
+  /** @brief Offset array of node type. */
+  torch::Tensor node_type_offset;
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:54]: Make sense, add for fix.

File: graphbolt/include/csc_sampling_graph.h (hunk 16 above)
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:58]: added

File: graphbolt/include/csc_sampling_graph.h (hunk 17 above)
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h]: fixed

File: graphbolt/include/csc_sampling_graph.h [hunk 20]
```diff
@@ -16,22 +16,43 @@ namespace sampling {
...
- * @brief Heterogeneous information in the Graph.
+ * @brief Structure representing heterogeneous information about a graph.
+ *
+ * Example usage:
+ *
+ * HeteroInfo info({"A", "B", "C"}, {"X", "Y", "Z"}, node_offset, edge_types);
+ *
+ * This example creates a `HeteroInfo` object with three node types ("A", "B",
+ * "C") and three edge types ("X", "Y", "Z"). The `node_offset` tensor
+ * represents the offset array of node type, and the `edge_types` tensor
```
Comment (frozenbugs) [inline graphbolt/include/csc_sampling_graph.h:27]: How does the `node_offset` tensor represents the offset array of node type?

Give an example?

Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h:27]: sure, add the example

File: graphbolt/include/csc_sampling_graph.h (hunk 13 above)
Comment (jermainewang) [reply graphbolt/include/csc_sampling_graph.h:129]: Then shall we check `num_rows_ == num_cols_` in this call? Otherwise, it will be ambiguous.

**Review by jermainewang — APPROVED** [review]

File: graphbolt/include/csc_sampling_graph.h (hunk 13 above)
Comment (peizhou001) [reply graphbolt/include/csc_sampling_graph.h]: I see, `NumNodes` is ambiguous if they are not same. Use `num_nodes `instead of `num_cols `and `num_rows` to make the it clear.

**Review by frozenbugs — APPROVED** [review]
